mlflow/store/artifact/cloud_artifact_repo.py

```diff
@@ -0,0 +1,252 @@
+                    paths=[
+                        staged_upload.artifact_file_path for staged_upload in staged_upload_chunk
+                    ],
+                )
+
+                inflight_uploads = {}
+                for staged_upload, write_credential_info in zip(
+                    staged_upload_chunk, write_credential_infos
+                ):
+                    upload_future = self.thread_pool.submit(
+                        self._upload_to_cloud,
+                        cloud_credential_info=write_credential_info,
+                        src_file_path=staged_upload.src_file_path,
+                        artifact_file_path=staged_upload.artifact_file_path,
+                    )
+                    inflight_uploads[staged_upload.src_file_path] = upload_future
+
+                yield from inflight_uploads.items()
+
+        with ArtifactProgressBar.files(
+            desc="Uploading artifacts", total=len(staged_uploads)
+        ) as pbar:
+            if len(staged_uploads) >= 10 and pbar.pbar:
+                _logger.info(
+                    "The progress bar can be disabled by setting the environment "
+                    f"variable {MLFLOW_ENABLE_ARTIFACTS_PROGRESS_BAR} to false"
+                )
+            for src_file_path, upload_future in upload_artifacts_iter():
+                try:
+                    upload_future.result()
+                    pbar.update()
+                except Exception as e:
+                    failed_uploads[src_file_path] = repr(e)
+
+        if len(failed_uploads) > 0:
+            raise MlflowException(
+                message=(
+                    "The following failures occurred while uploading one or more artifacts"
+                    f" to {self.artifact_uri}: {failed_uploads}"
+                )
+            )
+
+    @abstractmethod
+    def _get_write_credential_infos(self, paths):
+        """
+        For a batch of local files, get the write credentials for each file, which include
+        a presigned URL per file
+
+        Return a list of CredentialInfo objects, one for each file.
+        """
+        pass
+
+    @abstractmethod
+    def _upload_to_cloud(self, cloud_credential_info, src_file_path, artifact_file_path):
+        """
+        Upload a single file to the cloud.
+        :param cloud_credential_info: ArtifactCredentialInfo containing presigned URL for the
+                                      current file
+        :param src_file_path: source file path
+        :param artifact_file_path: artifact path, including file name
+        :return:
+        """
+        pass
+
+    # Read APIs
+
+    def _extract_headers_from_credentials(self, headers):
+        """
+        :return: A python dictionary of http headers converted from the protobuf credentials
+        """
+        return {header.name: header.value for header in headers}
+
+    def _parallelized_download_from_cloud(self, file_size, remote_file_path, local_path):
```

Comment:
This is an exact copy of `DatabricksArtifactRepository._parallelized_download_from_cloud()` except it inlines the read credential generation. This is because for non-parallel downloads (i.e. implementations of `_download_from_cloud`), not all subclasses need to generated read credentials. They can instead directly use cloud provider clients.